File: problem_0051.py
#!/usr/bin/env python3
"""
By replacing the 1st digit of the 2-digit number *3, it turns out that six of the nine possible values: 13, 23, 43, 53, 73, and 83, are all prime.

By replacing the 3rd and 4th digits of 56**3 with the same digit, this 5-digit number is the first example having seven primes among the ten generated numbers, yielding the family: 56003, 56113, 56333, 56443, 56663, 56773, and 56993. Consequently 56003, being the first member of this family, is the smallest prime with this property.

Find the smallest prime which, by replacing part of the number (not necessarily adjacent digits) with the same digit, is part of an eight prime value family.

"""

import logging
import re
import sys
import time

logging.basicConfig(
    stream=sys.stderr,
    level=logging.INFO,
    format='%(levelname)s: %(message)s')

def main():
    pattern_dict = {}
    for prime in prime_generator():
        for pattern in pattern_generator(str(prime)):
            if pattern in pattern_dict:
                pattern_dict[pattern].append(prime)
            else:
                pattern_dict[pattern] = [prime]
            if len(pattern_dict[pattern]) >= 8:
                print('Answer: {}'.format(pattern_dict[pattern][0]))
                return

def pattern_generator(string):
    unique_digits = list(set(string))
    for digit in unique_digits:
        temp =  re.sub(digit,'.',string)
        yield temp
        if temp.count('.') > 1:
            for pattern in sub_pattern(temp,digit):
                yield pattern

def sub_pattern(string,c):
    string = list(string)
    while True:
        for i in range(len(string)):
            if string[i] == c: string[i] = '.'
            elif string[i] == '.':
                string[i] = c
                break
        if '.' in string: yield ''.join(string)
        else: break

def prime_generator(block=100000):
    primes = []
    i = 0
    while True:
        if len(primes) <= i:
            primes = prime_sieve(block, primes)
        yield primes[i]
        i += 1

def prime_sieve(extend,primes=[]):
    logging.debug('Called prime_sieve')
    nums = [ True ] * (extend)
    if not primes: offset = 2
    else:
        offset = primes[-1]+1
        for p in primes:
            start = ( offset - 1 ) // p + 1
            end   = ( offset + extend - 1 ) // p + 1
            for n in range(start, end):
                logging.debug('{} is not prime ({} * {})'.format(n*p, n,p))
                nums[n*p-offset] = False
                logging.debug('Marked {} False'.format(n*p-offset))
    for i in range(extend):
        if nums[i]:
            logging.debug('Found {}'.format(i+offset))
            primes.append(i+offset)
            for n in range(2,1+extend//(i+offset)):
                logging.debug('{} is not prime it {} * {}'.format(
                    (i+offset)*n, n, i+offset))
                nums[(i+offset)*n-offset] = False
    return primes

if __name__ == "__main__":
    start = time.time()
    main()
    logging.info('That took {:4.2f} seconds'.format(time.time()-start))

File: problem_0051.hs
{-# LANGUAGE DeriveDataTypeable, RecordWildCards #-}

import Test.HUnit ((@=?), assertBool, runTestTT, Test(..))
import Text.Printf (printf)
import System.Console.CmdArgs
import Data.Time (getCurrentTime, diffUTCTime)

import Data.List (nub)
import qualified Data.Map as M

data EulerArgs =
    Euler
    | AdHoc{ len::Int }
    | UnitTest
    | RollUp{ start::Int, size::Int }
    deriving (Show, Data, Typeable)

-- By replacing the 1st digit of the 2-digit number *3, it turns out that six of the nine possible values: 13, 23, 43, 53, 73, and 83, are all prime.
-- By replacing the 3rd and 4th digits of 56**3 with the same digit, this 5-digit number is the first example having seven primes among the ten generated numbers, yielding the family: 56003, 56113, 56333, 56443, 56663, 56773, and 56993. Consequently 56003, being the first member of this family, is the smallest prime with this property.
-- Find the smallest prime which, by replacing part of the number (not necessarily adjacent digits) with the same digit, is part of an eight prime value family.

problem0051 :: Int -> [Integer]
problem0051 l = head $ filter ((==l) . length) $ rollUp M.empty primePatterns 
    where
        primePatterns = [(pat, p) | p <- primes, pat <- (patters '.' $ show p)]

rollUp :: (Ord a) => M.Map a [b] -> [(a, b)] -> [[b]]
rollUp _ [] = []
rollUp cache ((k, v):vks) = vs : (rollUp cache' vks)
    where
        vs = v : M.findWithDefault [] k cache
        cache' = M.insertWith (++) k [v] cache

rollUpTest = [
    [] @=? rollUp M.empty ([]::[(Int,Int)]),
    [[1],[2],[3,1],[4,2]] @=? rollUp M.empty [(1,1),(0,2),(1,3),(0,4)]]

substitue :: (Eq a) => a -> a -> [a] -> [[a]]
substitue _ _ [] = [[]]
substitue char r (t:text)
    | t /= char = map ((:)t) remainer
    | t == char = (map ((:)t) remainer) ++ (map ((:)r) remainer)
    where
        remainer = substitue char r text

substitueTest = [
    [""] @=? substitue 'a' '.' "",
    ["11", "1.", ".1", ".."] @=? substitue '1' '.' "11",
    ["aba","ab.",".ba",".b."] @=? substitue 'a' '.' "aba",
    ["a","."] @=? substitue 'a' '.' "a"]

patters :: (Eq a) => a -> [a] -> [[a]]
patters marker text = tail $ nub $ concat [substitue i marker text | i <- nub text]

pattersTest = [
    ["."] @=? patters '.' "1",
    ["1.",".1",".."] @=? patters '.' "11",
    ["12.",".21",".2.","1.1"] @=? patters '.' "121",
    ["123.",".231",".23.","1.31","12.1"] @=? patters '.' "1231"]

primes :: [Integer]
primes = 2 : [n | n <- [3,5..], isPrime n]

primesTest = [
    [2, 3, 5, 7, 11, 13, 17, 19, 23, 29] @=? (take 10 primes),
    [2, 3] @=? (take 2 primes),
    2 @=? (head primes)]

isPrime :: Integer -> Bool
isPrime 0 = False
isPrime 1 = False
isPrime n = all noRemainer primes'
    where
        primes' = takeWhile (\p -> p * p <= n) primes
        noRemainer d = 0 /= rem n d

isPrimeTest = [
    assertBool "9 is not prime" (not $ isPrime 9),
    assertBool "6 is not prime" (not $ isPrime 6),
    assertBool "4 is not prime" (not $ isPrime 4),
    assertBool "0 is not prime" (not $ isPrime 0),
    assertBool "3 is prime" (isPrime 3),
    assertBool "2 is prime" (isPrime 2)]

unitTests = map TestCase $
    isPrimeTest ++
    rollUpTest ++
    substitueTest ++
    pattersTest

exec :: EulerArgs -> IO ()
exec Euler = do
    let  answer = problem0051 8
    printf "Answer: %d\n" (last answer)
exec AdHoc{..} = do
    let answer = problem0051 len
    printf "Answer: %d\n" (last answer)
    print answer
exec UnitTest = do
    runTestTT $ TestList unitTests
    return ()
exec RollUp{..}= do
    let sets = take size $ drop start [(pat, p) | p <- primes, pat <- (patters '.' $ show p)]
    mapM_ print sets

main :: IO ()
main = do
    args <- cmdArgs $ modes [
        Euler,
        AdHoc{ len=7 },
        UnitTest,
        RollUp{ start = 0, size = 10 }]
    start <- getCurrentTime
    exec args
    stop <- getCurrentTime
    printf "That took %s\n"  $ show $ diffUTCTime stop start
